### PR TITLE
feat: add traffic shaper for download tasks

### PR DESCRIPTION
**Pull request by greenhandatsjtu**
Signed-off-by: greenhandatsjtu <sunhengke@sjtu.edu.cn>

## Description

For now, when a peer download from seed peers, the speed is limited by `perPeerRateLimit` and `totalRateLimit`. 
However, when there's only one task running with `perPeerRateLimit` set to 20M and `totalRateLimit` set to 100M, then 80% of the bandwidth is wasted. That's why we introduce traffic shaper. 
With the help of sampling traffic shaper, the rate limit of every task can be dynamically adjusted based on the context, the download speed of all tasks in the past second, which significantly improves the download performance.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: client/daemon/peer/traffic_shaper.go [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+		te.usedBandwidth.Store(0)
+	}
+
+	// allocate delta bandwidth for tasks
+	for _, te := range ts.tasks {
+		var deltaLimit float64
+		deltaLimit = (float64(ts.totalRateLimit) - float64(totalLeastBandwidth)) * (float64(te.needBandwidth) / float64(totalNeedBandwidth))
+		// make sure new limit is not smaller than pieceSize
+		if deltaLimit < 0 {
+			deltaLimit = 0
```
Comment (jim3ma) [inline client/daemon/peer/traffic_shaper.go:206]: If deltaLimit == 0, can we did not call `te.ptc.limiter.SetLimit`?

File: client/daemon/peer/traffic_shaper.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+func (ts *samplingTrafficShaper) AddTask(taskID string, ptc *peerTaskConductor) {
+	ts.Lock()
+	defer ts.Unlock()
+	nTasks := len(ts.tasks)
+	if nTasks == 0 {
+		nTasks++
+	}
+	limit := rate.Limit(float64(ts.totalRateLimit) / float64(nTasks))
```
Comment (jim3ma) [inline client/daemon/peer/traffic_shaper.go:220]: Check whether `limit` is less than `pieceSize`. We can use `math.Max(float64(pieceSize), float64(ts.totalRateLimit) / float64(nTasks))`.

**Review by jim3ma — CHANGES_REQUESTED** [review]

File: client/daemon/peer/traffic_shaper.go (hunk 1 above)
Comment (greenhandatsjtu) [reply client/daemon/peer/traffic_shaper.go]: `deltaLimit` indicates the difference between the allocated bandwidth and pieceSize, so if `deltaLimit` == 0, we still have to update the limit. The variable name is confusing, I have changed it to `diffLimit`

**Review by jim3ma — APPROVED** [review]

**Review by gaius-qi — APPROVED** [review]
LGTM

**Pull request merged**